### PR TITLE
Do not steal focus back to the cell when editor is closing by clicking outside the grid

### PR DESCRIPTION
**Pull request by amanmahajan7**

### CODE REVIEW

File: src/hooks/useRovingCellRef.ts [hunk 1]
```diff
@@ -14,6 +14,14 @@ export function useRovingCellRef(isSelected: boolean) {
       return;
     }
 
+    if (
```
Comment (amanmahajan7) [inline src/hooks/useRovingCellRef.ts:17]: Need to do some testing but seems like it is working unless I missed something

File: src/hooks/useRovingCellRef.ts [hunk 2]
```diff
@@ -2,32 +2,41 @@ import { useCallback, useState } from 'react';
...
-    if (cell.contains(document.activeElement)) return;
-    cell.focus({ preventScroll: true });
-  }, []);
+  const ref = useCallback(
+    (cell: HTMLDivElement | null) => {
+      if (cell === null) return;
+      scrollIntoView(cell);
+      if (
+        cell.contains(document.activeElement) ||
+        (isModeChangedFromEditToSelect && !cell.closest('.rdg')!.contains(document.activeElement))
```
Comment (amanmahajan7) [inline src/hooks/useRovingCellRef.ts:19]: The idea is to check if the active element is outside the grid when the editor is closed. AFAICT `isModeChangedFromEditToSelect` is not needed and grid does not steal focus but some tests fail without it and I am not sure why. It also seems safer to keep it.

File: website/demos/CommonFeatures.tsx [hunk 3]
```diff
@@ -377,6 +377,7 @@ export default function CommonFeatures({ direction }: Props) {
         <ExportButton onExport={() => exportToPdf(gridElement, 'CommonFeatures.pdf')}>
           Export to PDF
         </ExportButton>
+        <input type="text" />
```
Comment (amanmahajan7) [inline website/demos/CommonFeatures.tsx:380]: I will remove it before merging

Comment (amanmahajan7) [inline website/demos/CommonFeatures.tsx:380]: I will remove this before merge

File: src/EditCell.tsx [hunk 4]
```diff
@@ -53,10 +55,15 @@ export default function EditCell<R, SR>({
   });
 
   useEffect(() => {
-    if (!commitOnOutsideClick) return;
-
-    function onWindowCaptureMouseDown() {
-      frameRequestRef.current = requestAnimationFrame(commitOnOutsideMouseDown);
+    function onWindowCaptureMouseDown(event: MouseEvent) {
+      if (event.target instanceof Element) {
+        outsideClickedElementRef.current = event.target;
```
Comment (amanmahajan7) [inline src/EditCell.tsx:60]: Idea is to keep track of the clicked element and if it outside the grid then do not steal focus

File: src/hooks/useRovingCellRef.ts [hunk 5]
```diff
@@ -1,33 +1,41 @@
...
-    if (cell.contains(document.activeElement)) return;
-    cell.focus({ preventScroll: true });
-  }, []);
+  const ref = useCallback(
+    (cell: HTMLDivElement | null) => {
+      if (cell === null) return;
+      scrollIntoView(cell);
+      if (cell.contains(document.activeElement)) return;
+      if (skipCellFocus?.current) {
+        skipCellFocus.current = false;
```
Comment (amanmahajan7) [inline src/hooks/useRovingCellRef.ts:20]: Not sure if this is safe 🤔

Comment (amanmahajan7) [reply src/hooks/useRovingCellRef.ts:20]: It is safe.
https://reactjs.org/docs/hooks-faq.html#how-can-i-measure-a-dom-node

**Review by nstepien — COMMENTED** [review]
Can't review much more than this on a phone

File: src/EditCell.tsx [hunk 6]
```diff
@@ -53,10 +55,13 @@ export default function EditCell<R, SR>({
   });
 
   useEffect(() => {
-    if (!commitOnOutsideClick) return;
-
-    function onWindowCaptureMouseDown() {
-      frameRequestRef.current = requestAnimationFrame(commitOnOutsideMouseDown);
+    function onWindowCaptureMouseDown(event: MouseEvent) {
+      if (event.target instanceof Element) {
```
[the remark this answers is not in the document]
Comment (amanmahajan7) [reply src/EditCell.tsx]: Changed to Node

**Review by nstepien — COMMENTED** [review]
Changes look good, but I'm not 100% confident this won't lead to other edge-cases we haven't considered.

File: src/EditCell.tsx [hunk 7]
```diff
@@ -53,10 +55,13 @@ export default function EditCell<R, SR>({
   });
 
   useEffect(() => {
-    if (!commitOnOutsideClick) return;
-
-    function onWindowCaptureMouseDown() {
-      frameRequestRef.current = requestAnimationFrame(commitOnOutsideMouseDown);
+    function onWindowCaptureMouseDown(event: MouseEvent) {
+      if (event.target instanceof Node) {
```
[the remark this answers is not in the document]
Comment (amanmahajan7) [reply src/EditCell.tsx:59]: Getting this error
image

File: src/hooks/useRovingCellRef.ts [hunk 8]
```diff
@@ -1,33 +1,41 @@
...
+  );
 
   function onFocus(event: React.FocusEvent<HTMLDivElement>) {
     if (event.target !== event.currentTarget) {
       setIsChildFocused(true);
     }
   }
 
-  const isFocused = isSelected && !isChildFocused;
+  const isFocusable = isSelected && !isChildFocused;
```
[the remark this answers is not in the document]
Comment (amanmahajan7) [reply src/hooks/useRovingCellRef.ts:34]: We no longer focus so I changed the name. It is possible to tab back to this cell so I renamed it. Wdyt?

File: src/DataGrid.tsx [hunk 9]
```diff
@@ -979,8 +980,9 @@ function DataGrid<R, SR, K extends Key>(
       setSelectedPosition(({ idx, rowIdx }) => ({ idx, rowIdx, mode: 'SELECT' }));
     };
 
-    const onRowChange = (row: R, commitChanges?: boolean) => {
+    const onRowChange = (row: R, commitChanges?: boolean, clickedNode?: Node | null) => {
       if (commitChanges) {
+        skipCellFocus.current = clickedNode != null && !gridRef.current!.contains(clickedNode);
```
[the remark this answers is not in the document]
Comment (amanmahajan7) [reply src/DataGrid.tsx:985]: May be if the user click on something within the grid and scroll to remove the cells 🤔

File: src/EditCell.tsx [hunk 10]
```diff
@@ -30,15 +30,20 @@ const cellEditing = css`
 type SharedCellRendererProps<R, SR> = Pick<CellRendererProps<R, SR>, 'colSpan'>;
 
 interface EditCellProps<R, SR>
-  extends Omit<EditorProps<R, SR>, 'onClose'>,
+  extends Omit<EditorProps<R, SR>, 'onClose' | 'onRowChange'>,
     SharedCellRendererProps<R, SR> {
+  gridRef: RefObject<HTMLDivElement>;
+  skipCellFocusRef: MutableRefObject<boolean>;
+  onRowChange: (row: R, commitChanges?: boolean, clickedNode?: Node | null) => void;
```
Comment (nstepien) [inline src/EditCell.tsx:37]: Do we still need this change?

Comment (amanmahajan7) [reply src/EditCell.tsx]: Nope. Reverted

File: src/EditCell.tsx [hunk 11]
```diff
@@ -30,15 +30,20 @@ const cellEditing = css`
 type SharedCellRendererProps<R, SR> = Pick<CellRendererProps<R, SR>, 'colSpan'>;
 
 interface EditCellProps<R, SR>
-  extends Omit<EditorProps<R, SR>, 'onClose'>,
+  extends Omit<EditorProps<R, SR>, 'onClose' | 'onRowChange'>,
```
Comment (nstepien) [inline src/EditCell.tsx:33]: 🤔 
```suggestion
  extends Omit<EditorProps<R, SR>, 'onClose'>,
```

File: src/EditCell.tsx [hunk 12]
```diff
@@ -30,15 +30,20 @@ const cellEditing = css`
...
     SharedCellRendererProps<R, SR> {
+  gridRef: RefObject<HTMLDivElement>;
+  skipCellFocusRef: MutableRefObject<boolean>;
+  onRowChange: (row: R, commitChanges?: boolean) => void;
```
Comment (nstepien) [inline src/EditCell.tsx:37]:
```suggestion
```

**Review by nstepien — APPROVED** [review]

**Pull request merged**